src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -5280,6 +5280,59 @@ void C2_MacroAssembler::vector_mask_compress(KRegister dst, KRegister src, Regis
   kmov(dst, rtmp2);
 }
 
+#ifdef _LP64
+void C2_MacroAssembler::vector_compress_expand_avx2(int opcode, XMMRegister dst, XMMRegister src, XMMRegister mask,
+                                                    Register rtmp, Register rscratch, XMMRegister permv,
+                                                    XMMRegister xtmp, XMMRegister xtmp1, BasicType bt, int vec_enc) {
+  assert(type2aelembytes(bt) >= 4, "");
+  assert(opcode == Op_CompressV || opcode == Op_ExpandV, "");
+  if (bt == T_INT || bt == T_FLOAT) {
+    vmovmskps(rtmp, mask, vec_enc);
+    shlq(rtmp, 5);  // for 32 bit rows (8 ints)
+    if (opcode == Op_CompressV) {
+      lea(rscratch, ExternalAddress(StubRoutines::x86::compress_perm_table32()));
+    } else {
+      lea(rscratch, ExternalAddress(StubRoutines::x86::expand_perm_table32()));
+    }
+    addptr(rtmp, rscratch);
+    vmovdqu(permv, Address(rtmp));
+    vpermps(dst, permv, src, Assembler::AVX_256bit);
+    vpxor(xtmp, xtmp, xtmp, vec_enc);
+    // Blend the result with zero vector using permute mask, each row of
+    // permute table contains either a valid permute index or a -1 (default)
+    // value, this can potentially be used as a blending mask after
+    // compressing/expanding the source vector lanes.
+    vblendvps(dst, dst, xtmp, permv, vec_enc, false, xtmp1);
+  } else {
+    assert(bt == T_LONG || bt == T_DOUBLE, "");
+    vmovmskpd(rtmp, mask, vec_enc);
+    shlq(rtmp, 5); // for 64 bit rows (4 longs)
```

Comment:
```suggestion
    shlq(rtmp, 5); // for 32 bit rows (4 longs)
```

Reply:
Each long/double permute lane holds 64 bit value.

Reply:
@jatin-bhateja so why do you shift by 5? I thought 4 longs are 32 bit?

Reply:
For long/double each permute row is 32 byte in size, so a shift by 5 to compute row address.

Reply:
Ah right. Maybe we could say `32byte = 4 long = 4 * 64bit`.
Because "64bit row" sounds like the whole row is only 64 bit long. It is actually the cells that are 64bits, not the rows!